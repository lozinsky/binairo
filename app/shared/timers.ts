export interface Abortable {
  signal?: AbortSignal;
}

export async function setTimeout<T = void>(delay?: number, value?: T, options?: Abortable) {
  let resolve!: (value: T) => void;
  let reject!: (reason: unknown) => void;

  const args = ([] as T[]).concat(value ?? []);
  const signal = options?.signal;
  const promise = new Promise<T>((resolveCallback, rejectCallback) => {
    resolve = resolveCallback;
    reject = rejectCallback;
  });

  if (signal === undefined) {
    globalThis.setTimeout(resolve, delay, ...args);
    return await promise;
  }

  if (signal.aborted) {
    reject(signal.reason);
    return await promise;
  }

  const handleAbort = () => {
    globalThis.clearTimeout(handle);
    reject(signal.reason);
  };

  signal.addEventListener('abort', handleAbort);

  const handle = globalThis.setTimeout(resolve, delay, ...args);

  try {
    return await promise;
  } finally {
    signal.removeEventListener('abort', handleAbort);
  }
}
