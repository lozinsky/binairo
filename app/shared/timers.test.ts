/* eslint-disable @typescript-eslint/no-floating-promises, no-empty */

import { expect, test, vi } from 'vitest';

import { setTimeout } from './timers';

test('resolves on timeout if signal is not provided', () => {
  const promise = setTimeout(10);

  expect(promise).resolves.toMatchSnapshot();
});

test('resolves on timeout if signal is not provided with value', () => {
  const promise = setTimeout(10, 'value');

  expect(promise).resolves.toMatchSnapshot();
});

test('resolves on timeout if signal is provided', () => {
  const promise = setTimeout(10, undefined, {
    signal: new AbortController().signal,
  });

  expect(promise).resolves.toMatchSnapshot();
});

test('resolves on timeout if signal is provided with value', () => {
  const promise = setTimeout(10, 'value', {
    signal: new AbortController().signal,
  });

  expect(promise).resolves.toMatchSnapshot();
});

test('rejects if signal is aborted', () => {
  const controller = new AbortController();
  const promise = setTimeout(10, undefined, {
    signal: controller.signal,
  });

  controller.abort();

  expect(promise).rejects.toMatchSnapshot();
});

test('clears timeout if signal is aborted', async () => {
  const controller = new AbortController();
  const clearTimeoutMock = vi.spyOn(globalThis, 'clearTimeout');
  const promise = setTimeout(10, undefined, {
    signal: controller.signal,
  });

  controller.abort();

  try {
    await promise;
  } catch {}

  expect(clearTimeoutMock).toHaveBeenCalledOnce();
});

test('rejects if signal is already aborted', () => {
  const promise = setTimeout(10, undefined, {
    signal: AbortSignal.abort(),
  });

  expect(promise).rejects.toMatchSnapshot();
});

test('does not clear timeout if signal is already aborted', async () => {
  const clearTimeoutMock = vi.spyOn(globalThis, 'clearTimeout');
  const promise = setTimeout(10, undefined, {
    signal: AbortSignal.abort(),
  });

  try {
    await promise;
  } catch {}

  expect(clearTimeoutMock).not.toBeCalled();
});

test('does not set timeout if signal is already aborted', async () => {
  const setTimeoutMock = vi.spyOn(globalThis, 'setTimeout');
  const promise = setTimeout(10, undefined, {
    signal: AbortSignal.abort(),
  });

  try {
    await promise;
  } catch {}

  expect(setTimeoutMock).not.toBeCalled();
});
