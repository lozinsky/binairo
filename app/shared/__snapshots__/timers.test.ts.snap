// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rejects if signal is aborted 1`] = `[AbortError: This operation was aborted]`;

exports[`rejects if signal is already aborted 1`] = `[AbortError: This operation was aborted]`;

exports[`resolves on timeout if signal is not provided 1`] = `undefined`;

exports[`resolves on timeout if signal is not provided with value 1`] = `"value"`;

exports[`resolves on timeout if signal is provided 1`] = `undefined`;

exports[`resolves on timeout if signal is provided with value 1`] = `"value"`;
